### PR TITLE
Creating Nuget and Android packages for Training

### PR DESCRIPTION
**Pull request by askhade**
### Description
This PR creates Nuget and Android for Training. 

### Motivation and Context
These packages are intended to be released in ORT 1.15 to enable On-Device Training Scenarios.

### Nuget Packages:
1. New Native package -> **Microsoft.ML.OnnxRuntime.Training** (Native package will contain binaries for: win-x86, win-x64, win-arm, win-arm64, linux-x64, linux-arm64, android)
2. C# bindings will be added to existing package -> **Microsoft.ML.OnnxRuntime.Managed**

### Android Package published to Maven:
1. New package for training (full build) -> **onnxruntime-training-android-full-aar**

### Python Package published to PyPi:
1. Python bindings and offline tooling will be added to the existing ort training package -> **onnxruntime-training**

### CODE REVIEW

**Review by snnn — DISMISSED** [review]
LGTM. 

Thanks for the hard work!

File: csharp/src/Microsoft.ML.OnnxRuntime/Microsoft.ML.OnnxRuntime.csproj [hunk 1]
```diff
@@ -194,7 +203,7 @@
     <DefineConstants>$(DefineConstants);$(OrtConstants)</DefineConstants>
   </PropertyGroup>
 
-  <!-- debug output - makes finding/fixing any issues with the the conditions easy.
+  <!-- debug output - makes finding/fixing any issues with the the conditions easy.-->
```
Comment (askhade) [inline csharp/src/Microsoft.ML.OnnxRuntime/Microsoft.ML.OnnxRuntime.csproj:206]: will comment this section again

File: tools/ci_build/github/azure-pipelines/templates/linux-cpu-packaging-pipeline.yml [hunk 2]
```diff
@@ -4,8 +4,26 @@ parameters:
   type: string
   default: ''
 
+- name: stage_name_suffix
+  displayName: job name for nuget
```
Comment (edgchen1) [inline tools/ci_build/github/azure-pipelines/templates/linux-cpu-packaging-pipeline.yml:8]: I don't quite understand this description. looks like this value is part of the stage name.

File: tools/nuget/validate_package.py [hunk 3]
```diff
@@ -204,6 +224,7 @@ def validate_tarball(args):
         is_windows_ai_package,
         is_gpu_package,
         is_dml_package,
+        False,
```
Comment (edgchen1) [inline tools/nuget/validate_package.py:227]: nit: using a variable, e.g., `is_training_package = False` (as is done for `is_dml_package`) would be clearer than literal `False`

File: orttraining/orttraining/test/training_ops/cpu/tensor/gather_nd_grad_op_test.cc [hunk 4]
```diff
@@ -1,6 +1,10 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
+// TODO(askhade): enable these tests for all training builds.
```
Comment (baijumeswani) [inline orttraining/orttraining/test/training_ops/cpu/tensor/gather_nd_grad_op_test.cc:4]: Are tests failing at random? Shall we disable only those that are failing so we don't need to go looking for the failing ones when we work on fixing it?

File: tools/ci_build/github/azure-pipelines/nuget/templates/test_win.yml [hunk 5]
```diff
@@ -3,11 +3,14 @@ parameters:
   NugetPackageName : ''
   ArtifactSuffix: ''
   Skipx86Tests: 'false'
+  StageSuffix: 'CPU'
+  # default is for inference packages, for training packages use '-training'
+  ExtraArtifactNameSuffix: ''
```
[the remark this answers is not in the document]
Comment (baijumeswani) [reply tools/ci_build/github/azure-pipelines/nuget/templates/test_win.yml:8]: Seems like `ArtifactSuffix` is no longer used after this chance. Maybe we can repurpose  `ArtifactSuffix` to be what `ExtraArtifactNameSuffix` is?

File: java/build-android.gradle [hunk 6]
```diff
@@ -137,8 +144,8 @@ publishing {
 			artifact sourcesJar
 
 			pom {
-				name = 'onnx-runtime'
-				description = isMobileBuild ? mobileDescription : defaultDescription
+				name = enableTrainingApis == "1" ? 'onnxruntime-training' : 'onnx-runtime'
```
[the remark this answers is not in the document]
Comment (askhade) [reply java/build-android.gradle:147]: I thought so too. I did not change as I did not know the implication of this change.

**Review by baijumeswani — DISMISSED** [review]

**Review by snnn — APPROVED** [review]

**Review by edgchen1 — DISMISSED** [review]

File: tools/ci_build/github/android/build_aar_package.py [hunk 7]
```diff
@@ -180,6 +180,9 @@ def _build_aar(args):
         "-DminSdkVer=" + str(build_settings["android_min_sdk_version"]),
         "-DtargetSdkVer=" + str(build_settings["android_target_sdk_version"]),
         "-DbuildVariant=" + str(build_settings["build_variant"]),
+        "-DENABLE_TRAINING_APIS=1"
+        if "--enable_training_apis" in build_settings["build_params"]
+        else "-DENABLE_TRAINING_APIS=0",
```
Comment (skottmckay) [inline tools/ci_build/github/android/build_aar_package.py:185]:
```suggestion
        "-DENABLE_TRAINING_APIS=1" if "--enable_training_apis" in build_settings["build_params"]
                                   else "-DENABLE_TRAINING_APIS=0",
```

nit: hard to skim the names

File: tools/ci_build/github/azure-pipelines/templates/c-api-training-cpu.yml
Comment (skottmckay) [inline tools/ci_build/github/azure-pipelines/templates/c-api-training-cpu.yml:1]: nit: the filename doesn't really convey what this config does.

File: tools/ci_build/github/azure-pipelines/templates/win-ci.yml [hunk 8]
```diff
@@ -52,6 +57,11 @@ parameters:
   type: string
   default: 'onnxruntime'
 
+#for inference packages '', for training packages '-training'
+- name: extraArtifactNameSuffix
```
Comment (skottmckay) [inline tools/ci_build/github/azure-pipelines/templates/win-ci.yml:61]: nit: confusing to know how to use this when artifact_name_prefix has '-training in it'
> #for inference packages 'onnxruntime', for training packages 'onnxruntime-training'

and extraArtifactNameSuffix says pass '-training' in that as well
> #for inference packages '', for training packages '-training'

can we have one arg for the 'onnxruntime' and one for any optional suffix like '-training' and put together the right names with just that?

e.g. base_artifact_name = prefix + suffix, and apply the suffix to other things as needed?

File: tools/ci_build/github/windows/extract_nuget_files.ps1 [hunk 9]
```diff
@@ -52,10 +55,22 @@ elseif ($aars.Count -gt 1) {
...
+if (Test-Path -Path $nuget_artifacts_dir\onnxruntime-training-win-x64-*) {
+  $is_training_pipeline = $true
+  Write-Output "onnxruntime-training-win-x64-* dir exists. This is a training pipeline."
+}
+
+if ($is_training_pipeline) {
+  Copy-Item -Path $nuget_artifacts_dir\onnxruntime-training-win-x64-*\lib\* -Destination $Env:BUILD_BINARIESDIRECTORY\RelWithDebInfo\RelWithDebInfo
+}
+else {
+  Copy-Item -Path $nuget_artifacts_dir\onnxruntime-win-x64-*\lib\* -Destination $Env:BUILD_BINARIESDIRECTORY\RelWithDebInfo\RelWithDebInfo
```
Comment (skottmckay) [inline tools/ci_build/github/windows/extract_nuget_files.ps1:71]: nit: would be good to capture in a comment why we do these copies now that you've spent time figuring it out

File: tools/nuget/generate_nuspec_for_native_nuget.py [hunk 10]
```diff
@@ -286,7 +294,11 @@ def generate_metadata(line_list, args):
     generate_owners(metadata_list, "Microsoft")
     generate_description(metadata_list, args.package_name)
     generate_copyright(metadata_list, "\xc2\xa9 " + "Microsoft Corporation. All rights reserved.")
-    generate_tags(metadata_list, "ONNX ONNX Runtime Machine Learning")
+    generate_tags(
+        metadata_list, "ONNX ONNX Runtime Machine Learning"
+    ) if "Microsoft.ML.OnnxRuntime.Training." in args.package_name else generate_tags(
+        metadata_list, "ONNX ONNXRuntime-Training Learning-on-The-Edge On-Device-Training MachineLearning"
```
Comment (skottmckay) [inline tools/nuget/generate_nuspec_for_native_nuget.py:300]: nit: we should add 'native' to there as per nuget guidelines.

Also not sure that we need two 'ONNX' entries if the words aren't quoted.

File: tools/nuget/validate_package.py [hunk 11]
```diff
@@ -233,6 +255,7 @@ def validate_zip(args):
...
```
Comment (skottmckay) [inline tools/nuget/validate_package.py:258]: could/should this also use an 'is_training_package' variable for clarity?

File: tools/nuget/generate_nuspec_for_native_nuget.py (hunk 10 above)
Comment (skottmckay) [reply tools/nuget/generate_nuspec_for_native_nuget.py:300]: https://learn.microsoft.com/en-us/nuget/guides/native-packages

> Be sure to include native in the <tags> section of your .nuspec to help other developers find your package by searching on that tag.

Comment (askhade) [reply tools/nuget/generate_nuspec_for_native_nuget.py:300]: I added native, tags are space delimited: So I added ONNX and ONNXRuntime-Training

File: tools/ci_build/github/azure-pipelines/templates/win-ci.yml (hunk 8 above)
Comment (askhade) [reply tools/ci_build/github/azure-pipelines/templates/win-ci.yml:61]: Done!

File: tools/ci_build/github/android/build_aar_package.py (hunk 7 above)
Comment (askhade) [reply tools/ci_build/github/android/build_aar_package.py:185]: Will defer this to a different PR. Needs changes in lot of files.

**Review by baijumeswani — DISMISSED** [review]

**Review by baijumeswani — APPROVED** [review]

**Review by snnn — APPROVED** [review]

**Pull request merged**